src/smc-webapp/frame-editors/latex-editor/actions.ts

```diff
@@ -268,17 +270,36 @@ export class Actions extends BaseActions<LatexEditorState> {
 
   async run_build(time: number, force: boolean): Promise<void> {
     this.setState({ build_logs: Map() });
+
+    // for knitr related documents, we have to first build the derived tex file ...
     if (this.knitr) {
       await this.run_knitr(time, force);
       if (this.store.get("knitr_error")) return;
     }
-    await this.run_latex(time, force);
+    // update_pdf=false, because it is defered until the end
+    await this.run_latex(time, force, false);
+    // ... and then patch the synctex file to align the source line numberings
     if (this.knitr) {
       await this.run_patch_synctex(time, force);
     }
+
     const s = this.store.unsafe_getIn(["build_logs", "latex", "stdout"]);
-    if (typeof s == "string" && s.indexOf("sagetex.sty") != -1) {
-      await this.run_sagetex(time, force);
+    let update_pdf = true;
+    if (typeof s == "string") {
+      if (s.indexOf("sagetex.sty") != -1) {
+        update_pdf = false;
+        await this.run_sagetex(time, force);
+      }
+      if (s.indexOf("pythontex.sty") != -1 || s.indexOf("PythonTeX") != -1) {
+        update_pdf = false;
+        await this.run_pythontex(time, force);
```

Comment:
`run_sagetex` and `run_pythontex` either call `update_pdf` on their own (in case of not running it or error) or call `run_latex` again without the `false` (triggering `update_pdf`)